pennylane/devices/mini_qubit.py

```diff
@@ -0,0 +1,144 @@
+
+    Args:
+        tape (.QuantumTape): a representation of a circuit
+        seed (Any): A seed to use to control the generation of samples.
+
+    """
+    # 1) create the initial state
+    state = np.zeros(2 ** len(tape.wires))
+    state[0] = 1.0
+
+    # 2) apply all the operations
+    for op in tape.operations:
+        op_mat = op.matrix(wire_order=tape.wires)
+        state = op_mat @ state
+
+    # 3) perform measurements
+    # note that shots are pulled from the tape, not from the device
+    if tape.shots:
+        samples = sample_state(state, shots=tape.shots.total_shots, seed=seed)
+        results = tuple(mp.process_samples(samples, tape.wires) for mp in tape.measurements)
+    else:
+        # diagonalize measurement transform handled diagonalizing gates
+        # otherwise we would need to apply them
+        results = tuple(mp.process_state(state, tape.wires) for mp in tape.measurements)
+    return results[0] if len(tape.measurements) == 1 else results
+
+
+operations = frozenset({"PauliX", "PauliY", "PauliZ", "Hadamard", "CNOT", "CZ", "RX", "RY", "RZ"})
+
+
+def supports_operation(op: qml.operation.Operator) -> bool:
+    """This function used by preprocessing determines what operations
+    are natively supported by the device.
+
+    While in theory ``simulate`` can support any operation with a matrix, we limit the target
+    gateset for improved testing and reference purposes.
+
+    """
+    return getattr(op, "name", None) in operations
+
+
+@simulator_tracking  # update device.tracker with some relevant information
+@single_tape_support  # add support for device.execute(tape) in addition to device.execute((tape,))
+class MiniQubit(Device):
+    """A slimmed down numpy-based simulator for reference and testing purposes.
+
+    Args:
+        wires (int, Iterable[Number, str]): Number of wires present on the device, or iterable that
+            contains unique labels for the wires as numbers (i.e., ``[-1, 0, 2]``) or strings
+            (``['ancilla', 'q1', 'q2']``). Default ``None`` if not specified. While this device allows
+            for ``wires`` to be unspecified at construction time, other devices may make this argument
+            mandatory. Devices can also implement additional restrictions on the possible wires.
+        shots (int, Sequence[int], Sequence[Union[int, Sequence[int]]]): The default number of shots
+            to use in executions involving this device. Note that during execution, shots
+            are pulled from the circuit, not from the device.
+        seed (Union[str, None, int, array_like[int], SeedSequence, BitGenerator, Generator, jax.random.PRNGKey]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``. This is an optional
+            keyword argument added to follow recommend NumPy best practices. Other devices do not need
+            this parameter if it does not make sense for them.
+
+    """
+
+    name = "mini.qubit"
```

Comment:
Minor comment, but if you want to name it `reference.qubit`, it will help enforce that it is a reference device for future developers, and minimize users finding/using it :)

Reply:
@albi3ro any concerns with calling it `reference.qubit`? I also like `reference` more than `mini` - which could sound like it's for small scale simulations.